sklearn/cluster/_birch.py

```diff
@@ -604,7 +611,9 @@ def _fit(self, X, partial):
                 self.root_.append_subcluster(new_subcluster1)
                 self.root_.append_subcluster(new_subcluster2)
 
-        centroids = np.concatenate([leaf.centroids_ for leaf in self._get_leaves()])
+        centroids = np.concatenate(
+            [leaf.centroids_ for leaf in self._get_leaves()], dtype=X.dtype
```

Comment:
I think we can make the `centriods_` themselves float32 from the beginning. I think it would involve passing the dtype into `_CFNode` and making sure everything is float32.